dotnet/src/Connectors/Connectors.Memory.SqliteVec/SqliteServiceCollectionExtensions.cs

```diff
@@ -1,87 +1,194 @@
+    /// Registers a <see cref="SqliteVectorStore"/> as <see cref="VectorStore"/>, with the specified connection string and service lifetime.
     /// </summary>
-    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="VectorStore"/> on.</param>
-    /// <param name="connectionString">Connection string for <see cref="SqliteConnection"/>.</param>
-    /// <param name="options">Optional options to further configure the <see cref="VectorStore"/>.</param>
-    /// <param name="serviceId">An optional service id to use as the service key.</param>
-    /// <returns>Service collection.</returns>
+    /// <inheritdoc cref="AddKeyedSqliteVectorStore"/>
     public static IServiceCollection AddSqliteVectorStore(
         this IServiceCollection services,
-        string connectionString,
-        SqliteVectorStoreOptions? options = default,
-        string? serviceId = default)
-        => services.AddKeyedSingleton<VectorStore>(
-            serviceId,
-            (sp, _) => new SqliteVectorStore(connectionString, options ?? sp.GetService<SqliteVectorStoreOptions>() ?? new() { EmbeddingGenerator = sp.GetService<IEmbeddingGenerator>() }));
+        Func<IServiceProvider, string> connectionStringProvider,
+        Func<IServiceProvider, SqliteVectorStoreOptions>? optionsProvider = null,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
+        => AddKeyedSqliteVectorStore(services, serviceKey: null, connectionStringProvider, optionsProvider, lifetime);
+
+    /// <summary>
+    /// Registers a keyed <see cref="SqliteVectorStore"/> as <see cref="VectorStore"/>, with the specified connection string and service lifetime.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="VectorStore"/> on.</param>
+    /// <param name="serviceKey">The key with which to associate the vector store.</param>
+    /// <param name="connectionStringProvider">The connection string provider.</param>
+    /// <param name="optionsProvider">Options provider to further configure the vector store.</param>
+    /// <param name="lifetime">The service lifetime for the store. Defaults to <see cref="ServiceLifetime.Singleton"/>.</param>
+    /// <returns>The service collection.</returns>
+    public static IServiceCollection AddKeyedSqliteVectorStore(
+        this IServiceCollection services,
+        object? serviceKey,
+        Func<IServiceProvider, string> connectionStringProvider,
+        Func<IServiceProvider, SqliteVectorStoreOptions>? optionsProvider = null,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
+    {
+        Verify.NotNull(services);
+        Verify.NotNull(connectionStringProvider);
+
+        services.Add(new ServiceDescriptor(typeof(SqliteVectorStore), serviceKey, (sp, _) =>
+        {
+            var connectionString = connectionStringProvider(sp);
```

Comment:
nit: I still slightly prefer just inlining these variables and making everything a single lambda expression without a block - but not important.